crypto/rsa/rsa_gen.c

```diff
@@ -71,6 +71,188 @@ int RSA_generate_multi_prime_key(RSA *rsa, int bits, int primes,
     return rsa_keygen(rsa->libctx, rsa, bits, primes, e_value, cb, 0);
 }
 
+DEFINE_STACK_OF(BIGNUM)
+
+/*
+ * Given input values, q, p, n, d and e, derive the exponents 
+ * and coefficients for each prime in this key, placing the result
+ * on thier respective exps and coeffs stacks
+ */
+int rsa_multiprime_derive(RSA *rsa, int bits, int primes,
+                                 BIGNUM *e_value,
+                                 STACK_OF(BIGNUM) *factors, STACK_OF(BIGNUM) *exps,
+                                 STACK_OF(BIGNUM) *coeffs)
+{
+    STACK_OF(BIGNUM) *pplist = NULL, *pdlist = NULL;
+    BIGNUM *factor = NULL, *newpp = NULL, *newpd = NULL;
+    BIGNUM *dval = NULL, *newexp = NULL, *newcoeff = NULL;
+    BIGNUM *p = NULL, *q = NULL;
+    BIGNUM *r0 = NULL, *r1 = NULL, *r2 = NULL;
+    BN_CTX *ctx = NULL;
+    int i;
+    int ret = 0;
+
+
+    ctx = BN_CTX_new_ex(rsa->libctx);
+    if (!ctx)
+        goto err;
```

Comment:
I could, but I'm not sure why I wouldn't just pass through the common exit point here.  BN_CTX_end/BN_CTX_free both do NULL checks, so its safe.  It might be a bit faster, but unless there's a coding style point that requires it, I'm not sure its worth having a separate return point.

Reply:
Think it is in the coding style..

Reply:
https://www.openssl.org/policies/technical/coding-style.html
Chapter 7 suggests that centralizing your exit points is preferable, with the exception being that if there is no common cleanup to be done, you can just return.  While thats true here, this is the only point in that function where its possible to return without additional cleanup.  Given that all the cleanup operations do their own null checks, it seems more intuitive to me to make the err label the unilateral exit point

Reply:
We must be interpreting this differently.
 'If there is no cleanup needed then just return directly'.
 This is done in a lot of places in the code.

Reply:
This will be a pattern that you propagate  in all your PR's as well as code reviews, so it might be good to be on the same page :)

Reply:
I don't think we're interpreting it differently, we're on the same page as to meaning.  I think the difference I think comes in regarding the level of requirement.

Based on the conversation had in the OTC meeting a few weeks ago, as well as the email thread regarding style choices. It was my understanding that very few things in the style guide are interpreted as a 'must' requirement - i.e. they are largely suggestions for which we are able to use our best judgement regarding adherence (this specific case being an instance in which I felt it better to have a single exit point, rather than an additional return statement.  Not saying thats right or wrong, only that that was the choice I made at the time).  You're absolutely correct that many functions follow that guideline, but many also do not (RSA_verify_PKCS1_PSS_mgf1, ossl_rsa_multip_calc_product, rsa_pss_param_print, ossl_rsa_pss_params_create, etc to cite a few examples).

I don't have any issue with changing it, if we really do consider it a requirement, but I'm struggling to see a pattern for when we enforce it, vs when we let it go.